src/formats/jpeg.zig

```diff
@@ -0,0 +1,1084 @@
+            return error.App0MarkerDoesNotExist;
+        }
+
+        // Header length
+        _ = try reader.readIntBig(u16);
+
+        var identifier_buffer: [4]u8 = undefined;
+        _ = try reader.read(identifier_buffer[0..]);
+
+        if (!std.mem.eql(u8, identifier_buffer[0..], "JFIF")) {
+            return error.JfifIdentifierNotSet;
+        }
+
+        // NUL byte after JFIF
+        _ = try reader.readByte();
+
+        const jfifRevision = try reader.readIntBig(u16);
+        const densityUnit = @intToEnum(DensityUnit, try reader.readByte());
+        const xDensity = try reader.readIntBig(u16);
+        const yDensity = try reader.readIntBig(u16);
+
+        const thumbnailWidth = try reader.readByte();
+        const thumbnailHeight = try reader.readByte();
+
+        if (thumbnailWidth != 0 or thumbnailHeight != 0) {
+            return error.ThumbnailImagesUnsupported;
+        }
+
+        // Make sure there are no application markers after us.
+        if (((try reader.readIntBig(u16)) & 0xFFF0) == @enumToInt(Markers.Application0)) {
+            return error.ExtraneousApplicationMarker;
+        }
+
+        try seek_stream.seekBy(-2);
+
+        return JFIFHeader{
+            .jfifRevision = jfifRevision,
+            .densityUnit = densityUnit,
+            .xDensity = xDensity,
+            .yDensity = yDensity,
+        };
+    }
+};
+
+const QuantizationTable = union(enum) {
+    Q8: [64]u8,
+    Q16: [64]u16,
+
+    pub fn init(precision: u8, reader: ImageReader) !QuantizationTable {
+        // 0 = 8 bits, 1 = 16 bits
+        switch (precision) {
+            0 => {
+                var table = QuantizationTable{ .Q8 = undefined };
+
+                var offset: usize = 0;
+                while (offset < 64) : (offset += 1) {
+                    const value = try reader.readByte();
+                    table.Q8[ZigzagOffsets[offset]] = value;
+                }
+
+                if (JPEG_DEBUG) {
+                    var i: usize = 0;
+                    while (i < 8) : (i += 1) {
+                        var j: usize = 0;
+                        while (j < 8) : (j += 1) {
+                            std.debug.print("{d:4} ", .{table.Q8[i * 8 + j]});
+                        }
+                        std.debug.print("\n", .{});
+                    }
+                }
+
+                return table;
+            },
+            1 => {
+                var table = QuantizationTable{ .Q16 = undefined };
+
+                var offset: usize = 0;
+                while (offset < 64) : (offset += 1) {
+                    const value = try reader.readIntBig(u16);
+                    table.Q16[ZigzagOffsets[offset]] = value;
+                }
+
+                return table;
+            },
+            else => return error.UnknownQuantizationTablePrecision,
+        }
+    }
+};
+
+const HuffmanCode = struct { length_minus_one: u4, code: u16 };
```

Comment:
Try looking at the huffman code in https://github.com/ziglang/zig/blob/0fb26b03690eb327164bbc7a21ba2d69d3b1a5b8/lib/std/compress/deflate.zig#L24-L134 to see if you could reuse and avoid using a dynamically allocated HashMap.

If it is too complicated, ignore this lol

Reply:
I'll have to take a look at that later, definitely sounds promising though. Shame it's not a `pub const`.